cilium: distributed lru map sizing roundup fixes

borkmann opened the pull request:
(see commit msg)

Closes: https://github.com/cilium/cilium/issues/39009

```release-note
Fix map recreation loop when distributed lru setting is enabled
```

borkmann committed:
cilium, maps: Fix distributed lru map sizing

When BPFDistributedLRU is set, the kernel internally adjusts the map's
maximum entry size as follows:

	htab->map.max_entries = roundup(attr->max_entries,
					num_possible_cpus());
	if (htab->map.max_entries < attr->max_entries)
		htab->map.max_entries = rounddown(attr->max_entries,
						  num_possible_cpus());

The problem with this kernel logic is that the next time Cilium tries
to regenerate endpoints, it will find that the map properties mismatch
from Cilium's request, since the original max_entries from Cilium has
been rounded up internally. So we keep continuously rebuilding all CT
and NAT maps, leading to connectivity disruption. The only way to solve
it is to adjust the Cilium's max_entries request also as a rounded up
value, only then we keep matching the map properties.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, maps: Only allow distributed lru with dynamic map sizing

The distributed lru setting makes sense with dynamic map sizing given
the kernel internally adjusts the map's max_entries as a rounded up
value to a multiple of num_possible_cpus().

With statically configured map sizing it can happen that the exact
sizing "breaks" in case of heterogenious nodes (w/ different # of CPUs)
in the cluster if the max CT/NAT map sizes were fixed to a specific
value.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

tommyp1ckles commented on pkg/option/config.go:
```diff
@@ -3655,13 +3660,26 @@ func (c *DaemonConfig) calculateDynamicBPFMapSizes(vp *viper.Viper, totalMemory
+		entries = roundUp(entries, possibleCPUs)
 		if entries < min {
-			entries = min
+			entries = roundUp(min, possibleCPUs)
```
maybe lets add a note on why we need this rounding logic (i.e. https://elixir.bootlin.com/linux/v6.13.7/source/kernel/bpf/hashtab.c#L496-L506)

borkmann committed:
cilium, tests: Add a test case for distributed LRU

Add a test case with distributed LRU and dynamic map sizing to make sure
we get CPU-aligned map sizing.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

tklauser commented on pkg/option/config.go:
```diff
@@ -45,6 +45,8 @@ import (
 	"github.com/cilium/cilium/pkg/time"
 	"github.com/cilium/cilium/pkg/version"
+
+	"github.com/cilium/ebpf"
```
nit: I think this import clause should be part of the 2nd import group (external non-stdlib dependencies) above. We generally try to stick to 3 import groups:

1. Go stdlib
2. External non-stdlib dependencies
3. Own module (github.com/cilium/cilium)

borkmann replied:
thanks, fixed now!

tklauser commented on pkg/option/config_test.go:
```diff
@@ -21,6 +21,8 @@ import (
 	"github.com/cilium/cilium/pkg/defaults"
 	ipamOption "github.com/cilium/cilium/pkg/ipam/option"
+
+	"github.com/cilium/ebpf"
```
Same here, this should be part of the 2nd import group.

borkmann committed:
cilium, docs: Add a small note about distributed lru requirements

Add a note that it only works on combination with dynamic map sizing.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

ti-mo reviewed (approved):
Thanks! As a minor nit, I'd just make roundUp/Down plain functions so they don't need to be redeclared in tests.

borkmann committed:
cilium, misc: Migrate roundUp and roundDown into util package

Move the implementation into its own util package for reuse, so we do not
have to open-code it each time.

Suggested-by: Timo Beckers <timo@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

aditighag reviewed (approved):
Nice find! Looks good for my CODEOWNERS.
